crates/nu-cli/src/completions/completer.rs

```diff
@@ -56,21 +57,95 @@ impl NuCompleter {
     }
 
+    fn external_completion(
+        &self,
+        block_id: usize,
+        spans: Vec<String>,
+        offset: usize,
+        span: Span,
+    ) -> Vec<Suggestion> {
+        let stack = self.stack.clone();
+        let block = self.engine_state.get_block(block_id);
+        let mut callee_stack = stack.gather_captures(&block.captures);
+
+        // Line
+        if let Some(pos_arg) = block.signature.required_positional.get(0) {
+            if let Some(var_id) = pos_arg.var_id {
+                callee_stack.add_var(
+                    var_id,
+                    Value::List {
+                        vals: spans
+                            .into_iter()
+                            .map(|it| Value::String {
+                                val: it,
+                                span: Span::unknown(),
+                            })
+                            .collect(),
+                        span: Span::unknown(),
+                    },
+                );
+            }
+        }
+
+        let result = eval_block(
+            &self.engine_state,
+            &mut callee_stack,
+            block,
+            PipelineData::new(span),
+            true,
+            true,
+        );
+
+        match result {
+            Ok(pd) => {
+                let value = pd.into_value(span);
+                if let Value::List { vals, span: _ } = value {
+                    let result = map_value_completions(
+                        vals.iter(),
+                        Span {
+                            start: span.start,
+                            end: span.end,
+                        },
+                        offset,
+                    );
+
+                    return result;
+                }
+            }
+            Err(err) => println!("failed to eval completer block: {}", err),
+        }
+
+        vec![]
+    }
+
     fn completion_helper(&mut self, line: &str, pos: usize) -> Vec<Suggestion> {
         let mut working_set = StateWorkingSet::new(&self.engine_state);
+        let original_pos = pos;
         let offset = working_set.next_span_start();
         let (mut new_line, alias_offset) = try_find_alias(line.as_bytes(), &working_set);
         let initial_line = line.to_string();
         new_line.push(b'a');
         let pos = offset + pos;
+        let config = self.engine_state.get_config();
+
         let (output, _err) = parse(&mut working_set, Some("completer"), &new_line, false, &[]);
 
         for pipeline in output.pipelines.into_iter() {
             for expr in pipeline.expressions {
                 let flattened: Vec<_> = flatten_expression(&working_set, &expr);
                 let span_offset: usize = alias_offset.iter().sum();
+                let mut spans: Vec<String> = vec![];
 
                 for (flat_idx, flat) in flattened.iter().enumerate() {
+                    // Skip the last 'a' as span item
+                    if flat_idx != flattened.len() {
```

Comment:
Just tried the new version and this didn't work for me (still had the 'a' suffix).

This seems to work:
```rust
                    let current_span_str = String::from_utf8_lossy(&current_span);
                    // Skip the last 'a' as span item
                    if flat_idx == flattened.len() - 1 {
                        let mut chars = current_span_str.chars();
                        chars.next_back();
                        let current_span_str = chars.as_str().to_owned();
                        spans.push(current_span_str.to_string());
                    } else {
                        spans.push(current_span_str.to_string());
                    }

```

```sh
let external_completer = {|spans|
   carapace $spans.0 nushell $spans | from json
}
```

Reply:
Open quotes aren't handled though.
```sh
chown ""<TAB> # works ["chown", ""]
chown "<TAB>  # fails ["chown", "\""]
```

Reply:
Yeah for the `'a'` I forgot to add the ` -1`

Reply:
> Open quotes aren't handled though.
> 
> ```shell
> chown ""<TAB> # works ["chown", ""]
> chown "<TAB>  # fails ["chown", "\""]
> ```

I added the issue in the description so I won't forget